### PR TITLE
feat(function): support dateAdd function

### LINKED ISSUE #5076: Combine datetime interval add&subsruct  to one DATE_ADD
**Summary**

Syntax:
```
​​DATE_ADD('<unit>', <interval>, <date_or_time_expr>)
```

`<unit>`: `SECOND`, `MINUTE`, `HOUR`, `DAY`, `WEEK`, `MONTH`, `YEAR`
`<interval>`: Number, this can be negative for substract
`<date_or_time_expr>`:  The expression evaluate to a date, time, or timestamp

Example:
```
SELECT toDate(18875), DATE_ADD('YEAR', 2, toDate(18875));
+---------------+-----------------------------+
| 2021-09-05    | 2023-09-05                  |
+---------------+-----------------------------+

SELECT toDate(18875), DATE_ADD('YEAR', -2, toDate(18875));
+---------------+-----------------------------+
| 2021-09-05    | 2019-09-05                  |
+---------------+-----------------------------+
```

### PR DESCRIPTION
**Pull request by fkuner**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support dateAdd function for new parser, for example:

```
select dateAdd(to_date(18321), 1, SECOND);
select dateAdd(to_date(18321), interval '1' YEAR);
```

The old parser don't support sql like `select dateAdd(to_date(18321), 1, SECOND);` but supports sql `select dateAdd(to_date(18321), interval '1' YEAR);
`

## Changelog

- New Feature

## Related Issues

Fixes #5076

### CODE REVIEW

File: common/functions/src/scalars/dates/date.rs [hunk 1]
```diff
@@ -124,5 +125,6 @@ impl DateFunction {
         factory.register("subtractHours", AddTimesFunction::desc(-3600));
         factory.register("subtractMinutes", AddTimesFunction::desc(-60));
         factory.register("subtractSeconds", AddTimesFunction::desc(-1));
+        factory.register("dateAdd", ArithmeticPlusFunction::desc());
```
Comment (sundy-li) [inline common/functions/src/scalars/dates/date.rs:128]: let's use `date_add`

**Comment by BohuTANG** [comment]
@mergify update

File: query/src/sql/planner/semantic/type_check.rs [hunk 2]
```diff
@@ -388,19 +389,34 @@ impl<'a> TypeChecker<'a> {
...
         let mut arg_types = vec![];
 
+        let mut suffix = String::new();
+
         for argument in arguments {
-            let (arg, arg_type) = self.resolve(argument, None).await?;
-            args.push(arg);
-            arg_types.push(arg_type);
+            match argument {
+                Expr::DateTimeUnit { unit, .. } => {
```
Comment (sundy-li) [inline query/src/sql/planner/semantic/type_check.rs:396]: interval expr UNIT -> to_interval()

Comment (sundy-li) [reply query/src/sql/planner/semantic/type_check.rs:396]: cast(a as UInt32) 


CastExpr {
  expr: Expr,
  data_type: DataTypeImpl
}

**Comment by fkuner** [comment]
The pr should be closed because of design problem.

**Comment by fkuner** [comment]
> > The pr should be closed because of design problem.
> 
> There is no need to close a PR unless we stop working on this. I think it's enough to Convert to draft and do a force push.

Yes, we stop working on this and I open a new pr fow the issue.

**Pull request closed without merging**